WalletWasabi/WabiSabi/Backend/Banning/CoinVerifier.cs

```diff
@@ -21,87 +19,84 @@ public CoinVerifier(CoinJoinIdStore coinJoinIdStore, CoinVerifierApiClient apiCl
 
-		var lastChangeId = Whitelist.ChangeId;
-		Whitelist.RemoveAllExpired(WabiSabiConfig.ReleaseFromWhitelistAfter);
-
-		var scriptsToCheck = new HashSet<Script>();
-		var innocentsCounter = 0;
+		// Booting up the results with the default value - ban: no, remove: yes.
+		Dictionary<Coin, CoinVerifyResult> coinVerifyItems = coinsToCheck.ToDictionary(c => c, c => new CoinVerifyResult(c, false, true));
 
+		List<Task<CoinVerifyResult>> tasks = new();
 		foreach (var coin in coinsToCheck)
 		{
-			if (CheckIfAlreadyVerified(coin))
-			{
-				innocentsCounter++;
-				yield return new CoinVerifyInfo(false, coin);
-			}
-			else
+			if (!CoinVerifyItems.TryGetValue(coin, out var item))
 			{
-				scriptsToCheck.Add(coin.ScriptPubKey);
+				// Quickly re-scheduling the missing items.
+				ScheduleVerification(coin, cancellationToken, TimeSpan.Zero);
+				if (!CoinVerifyItems.TryGetValue(coin, out item))
+				{
+					// This should not happen.
+					Logger.LogError("Coin cannot be re-scheduled for verification.");
+				}
 			}
+
+			tasks.Add(item.TaskCompletionSource.Task);
 		}
 
-		Logger.LogInfo($"{innocentsCounter} out of {coinsToCheck.Count()} utxo is already verified in Round({roundId}).");
-		await foreach (var response in CoinVerifierApiClient.VerifyScriptsAsync(scriptsToCheck, linkedCts.Token))
+		try
 		{
-			bool shouldBanUtxo = CheckForFlags(response.ApiResponseItem);
-
-			// Find all coins with the same script (address reuse).
-			foreach (var coin in coinsToCheck.Where(c => c.ScriptPubKey == response.ScriptPubKey))
+			do
 			{
-				if (!shouldBanUtxo)
-				{
-					Whitelist.Add(coin.Outpoint);
-				}
+				var completedTask = await Task.WhenAny(tasks).WaitAsync(linkedCts.Token).ConfigureAwait(false);
+				tasks.Remove(completedTask);
+				var result = await completedTask.WaitAsync(linkedCts.Token).ConfigureAwait(false);
+
+				// The verification task fulfilled its purpose.
+				CoinVerifyItems.TryRemove(result.Coin, out var _);
 
-				yield return new CoinVerifyInfo(shouldBanUtxo, coin);
+				// Update the default value with the real result.
+				coinVerifyItems[result.Coin] = result;
 			}
+			while (tasks.Any());
 		}
+		catch (OperationCanceledException ex)
+		{
+			if (cancellationTokenSource.IsCancellationRequested)
+			{
+				Logger.LogError(ex);
+			}
 
-		if (Whitelist.ChangeId != lastChangeId)
+			// Otherwise just return.
+		}
+		catch (Exception ex)
 		{
-			Whitelist.WriteToFile();
+			Logger.LogError(ex);
 		}
 
-		var duration = DateTimeOffset.UtcNow - before;
-		RequestTimeStatista.Instance.Add("verifier-period", duration);
+		CleanUp();
+
+		return coinVerifyItems.Values.ToArray();
+	}
+
+	private void CleanUp()
+	{
+		var now = DateTimeOffset.UtcNow;
+		foreach (var (coin, item) in CoinVerifyItems)
+		{
+			if (item.TaskCompletionSource.Task.IsCompleted && now - item.ScheduleTime > TimeSpan.FromDays(2))
```

Comment:
`TimeSpan.FromDays(2)` FromDays?

```suggestion
			if (item.TaskCompletionSource.Task.IsCompleted && now - item.ScheduleTime > TimeSpan.FromMinutes(2))
```

Reply:
Yes, we should set this timeout far bigger than the input reg of any round. It is a sanity timeout.

Reply:
OK, I see.